ext/dom/html_document.c

```diff
@@ -0,0 +1,1047 @@
+	*buf_ref_ref = buf_ref;
+	return true;
+fail_oom:
+	*buf_ref_ref = buf_ref;
+	return false;
+}
+
+static bool dom_parse_decode_encode_step(lexbor_libxml2_bridge_parse_context *ctx, lxb_html_document_t *document, lxb_html_parser_t *parser, const lxb_char_t **buf_ref_ref, const lxb_char_t *buf_end, dom_decoding_encoding_ctx *decoding_encoding_ctx, size_t *tokenizer_error_offset, size_t *tree_error_offset)
+{
+	if (decoding_encoding_ctx->fast_path) {
+		return dom_decode_encode_fast_path(ctx, document, parser, buf_ref_ref, buf_end, decoding_encoding_ctx, tokenizer_error_offset, tree_error_offset);
+	} else {
+		return dom_decode_encode_slow_path(ctx, document, parser, buf_ref_ref, buf_end, decoding_encoding_ctx, tokenizer_error_offset, tree_error_offset);
+	}
+}
+
+static bool dom_parse_decode_encode_finish(lexbor_libxml2_bridge_parse_context *ctx, lxb_html_document_t *document, lxb_html_parser_t *parser, dom_decoding_encoding_ctx *decoding_encoding_ctx, size_t *tokenizer_error_offset, size_t *tree_error_offset)
+{
+	if (!decoding_encoding_ctx->fast_path) {
+		/* Fast path handles codepoints one by one, so this part is not applicable in that case */
+		(void) lxb_encoding_decode_finish(&decoding_encoding_ctx->decode);
+		size_t decoding_buffer_size = lxb_encoding_decode_buf_used(&decoding_encoding_ctx->decode);
+		if (decoding_buffer_size > 0) {
+			const lxb_codepoint_t *codepoints_ref = (const lxb_codepoint_t *) decoding_encoding_ctx->codepoints;
+			const lxb_codepoint_t *codepoints_end = codepoints_ref + decoding_buffer_size;
+			(void) decoding_encoding_ctx->encode_data->encode(&decoding_encoding_ctx->encode, &codepoints_ref, codepoints_end);
+			if (!dom_process_parse_chunk(ctx, document, parser, lxb_encoding_encode_buf_used(&decoding_encoding_ctx->encode), decoding_encoding_ctx->encoding_output, decoding_buffer_size, tokenizer_error_offset, tree_error_offset)) {
+				return false;
+			}
+		}
+	}
+	(void) lxb_encoding_encode_finish(&decoding_encoding_ctx->encode);
+	if (lxb_encoding_encode_buf_used(&decoding_encoding_ctx->encode) && !dom_process_parse_chunk(ctx, document, parser, lxb_encoding_encode_buf_used(&decoding_encoding_ctx->encode), decoding_encoding_ctx->encoding_output, lxb_encoding_decode_buf_used(&decoding_encoding_ctx->decode), tokenizer_error_offset, tree_error_offset)) {
+		return false;
+	}
+	return true;
+}
+
+static bool check_options_validity(zend_long options)
+{
+	const zend_long VALID_OPTIONS = XML_PARSE_NOERROR | XML_PARSE_COMPACT | HTML_PARSE_NOIMPLIED | DOM_HTML_NO_DEFAULT_NS;
+	if ((options & ~VALID_OPTIONS) != 0) {
+		zend_argument_value_error(2, "contains invalid flags (allowed flags: LIBXML_NOERROR, LIBXML_COMPACT, LIBXML_HTML_NOIMPLIED, DOM\\NO_DEFAULT_NS)");
+		return false;
+	}
+	return true;
+}
+
+PHP_METHOD(DOM_HTMLDocument, createEmpty)
+{
+	const char *encoding = "UTF-8";
+	size_t encoding_len = strlen("UTF-8");
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|s", &encoding, &encoding_len) == FAILURE) {
```

Comment:
```suggestion
	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|p", &encoding, &encoding_len) == FAILURE) {
```
Seems like not wanting a null byte is done in the method below?